File: lib/features/ai_engine/anomaly_detection/anomaly_detector.dart
import '../analytics/stats_engine.dart';
import '../../models/app_models.dart';

class AnomalyDetector {
  static List<TransactionModel> detectByIQR(List<TransactionModel> transactions) {
    if (transactions.length < 4) return [];
    
    List<double> amounts = transactions.map((t) => t.amount).toList();
    List<double> percentiles = StatsEngine.getPercentiles(amounts, [25, 75]);
    
    double q1 = percentiles[0];
    double q3 = percentiles[1];
    double iqr = q3 - q1;
    
    double lowerBound = q1 - 1.5 * iqr;
    double upperBound = q3 + 1.5 * iqr;
    
    return transactions.where((t) => t.amount < lowerBound || t.amount > upperBound).toList();
  }

  static List<TransactionModel> detectByZScore(List<TransactionModel> transactions, {double threshold = 3.0}) {
    if (transactions.length < 2) return [];
    
    List<double> amounts = transactions.map((t) => t.amount).toList();
    double mean = StatsEngine.mean(amounts);
    double stdDev = StatsEngine.standardDeviation(amounts);
    
    if (stdDev == 0) return [];

    return transactions.where((t) {
      double zScore = (t.amount - mean).abs() / stdDev;
      return zScore > threshold;
    }).toList();
  }
}

File: src/features/ai_engine/anomaly_detector.ts
import { Transaction } from "../../models/types";
import { StatsEngine } from "../ai_engine/stats_engine";

export class AnomalyDetector {
  static detectByIQR(transactions: Transaction[]): Transaction[] {
    if (transactions.length < 4) return [];

    const amounts = transactions.map((t) => t.amount);
    const [q1, q3] = StatsEngine.getPercentiles(amounts, [25, 75]);

    const iqr = q3 - q1;
    const lowerBound = q1 - 1.5 * iqr;
    const upperBound = q3 + 1.5 * iqr;

    return transactions.filter(
      (t) => t.amount < lowerBound || t.amount > upperBound
    );
  }

  static detectByZScore(
    transactions: Transaction[],
    threshold: number = 3.0
  ): Transaction[] {
    if (transactions.length < 2) return [];

    const amounts = transactions.map((t) => t.amount);
    const mean = StatsEngine.mean(amounts);
    const stdDev = StatsEngine.standardDeviation(amounts);

    if (stdDev === 0) return [];

    return transactions.filter((t) => {
      const zScore = Math.abs(t.amount - mean) / stdDev;
      return zScore > threshold;
    });
  }
}
